### ISSUE TITLE
GSEA 4.0 fails to Infinite values 

### ISSUE THREAD

**Issue by ACastanza** [GSEA-MSigDB/gsea-desktop#14]
When a dataset contains an infinite value or data that GSEA calculates to be an infinite rank, GSEA 4.0 does not exclude it from the mountain plot, and instead of rendering the plot a Java error is inserted into the report.

This may also affect NaN calculations (-Inf)

GSEA 3 correctly renders the plot, however, the exact method of handling in the prior release is unknown (set +- inf = min/max?)

**Comment by davideby** [comment]
Just a quick note on this: we had to update the JFreeChart version in the Java 11 update and either the library itself or the code changes required for it are certainly the cause here.

I don't have a fix yet but will need to look at those changes in more detail.

**Comment by davideby** [comment]
Accidentally closed during PR merge...

**Comment by bounlu** [comment]
Hi. I had the same issue with version 4.0.3:

```
Creating marker selection reports ...
27032    [ERROR ] - Trouble saving image
java.lang.IllegalArgumentException: Must be finite.
	at org.jfree.chart.axis.NumberTickUnitSource.getCeilingTickUnit(NumberTickUnitSource.java:114) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.axis.NumberAxis.selectVerticalAutoTickUnit(NumberAxis.java:896) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.axis.NumberAxis.selectAutoTickUnit(NumberAxis.java:828) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.axis.NumberAxis.refreshTicksVertical(NumberAxis.java:1049) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.axis.NumberAxis.refreshTicks(NumberAxis.java:929) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.axis.ValueAxis.reserveSpace(ValueAxis.java:806) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.plot.XYPlot.calculateRangeAxisSpace(XYPlot.java:3077) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.plot.XYPlot.calculateAxisSpace(XYPlot.java:2988) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.plot.XYPlot.draw(XYPlot.java:3132) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.JFreeChart.draw(JFreeChart.java:1286) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.JFreeChart.createBufferedImage(JFreeChart.java:1453) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.JFreeChart.createBufferedImage(JFreeChart.java:1434) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.ChartUtils.writeChartAsPNG(ChartUtils.java:184) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.ChartUtils.saveChartAsPNG(ChartUtils.java:310) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.ChartUtils.saveChartAsPNG(ChartUtils.java:285) ~[jfreechart-1.5.0.jar:?]
	at edu.mit.broad.genome.charts.XChartImpl.saveAsPNG(XChartImpl.java:114) ~[gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.api.PicFile.save(PicFile.java:207) ~[gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.api.PicFile.createIMG(PicFile.java:176) ~[gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.pages.HtmlPage.addChart(HtmlPage.java:206) [gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.MiscReports.createDatasetHeatMapAndCorrelationPlots(MiscReports.java:91) [gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.EnrichmentReports.createGseaLikeReport(EnrichmentReports.java:326) [gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.EnrichmentReports.createGseaLikeReport(EnrichmentReports.java:174) [gsea-minimal-4.0.3.jar:?]
	at xtools.gsea.AbstractGsea2Tool.execute_one_with_reporting(AbstractGsea2Tool.java:163) [gsea-minimal-4.0.3.jar:?]
	at xtools.gsea.Gsea.execute(Gsea.java:160) [gsea-minimal-4.0.3.jar:?]
	at xtools.api.AbstractTool.module_main(AbstractTool.java:434) [gsea-minimal-4.0.3.jar:?]
	at org.genepattern.modules.GseaWrapper.main(GseaWrapper.java:273) [gsea-minimal-4.0.3.jar:?]
	at xapps.gsea.CLI.main(CLI.java:29) [gsea-minimal-4.0.3.jar:?]
Creating FDR reports ...
27367    [ERROR ] - Trouble saving image
java.lang.IllegalArgumentException: Must be finite.
	at org.jfree.chart.axis.NumberTickUnitSource.getCeilingTickUnit(NumberTickUnitSource.java:114) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.axis.NumberAxis.selectVerticalAutoTickUnit(NumberAxis.java:896) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.axis.NumberAxis.selectAutoTickUnit(NumberAxis.java:828) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.axis.NumberAxis.refreshTicksVertical(NumberAxis.java:1049) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.axis.NumberAxis.refreshTicks(NumberAxis.java:929) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.axis.ValueAxis.reserveSpace(ValueAxis.java:806) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.plot.XYPlot.calculateRangeAxisSpace(XYPlot.java:3077) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.plot.CombinedDomainXYPlot.calculateAxisSpace(CombinedDomainXYPlot.java:447) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.plot.CombinedDomainXYPlot.draw(CombinedDomainXYPlot.java:482) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.JFreeChart.draw(JFreeChart.java:1286) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.JFreeChart.createBufferedImage(JFreeChart.java:1453) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.JFreeChart.createBufferedImage(JFreeChart.java:1434) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.ChartUtils.writeChartAsPNG(ChartUtils.java:184) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.ChartUtils.saveChartAsPNG(ChartUtils.java:310) ~[jfreechart-1.5.0.jar:?]
	at org.jfree.chart.ChartUtils.saveChartAsPNG(ChartUtils.java:285) ~[jfreechart-1.5.0.jar:?]
	at edu.mit.broad.genome.charts.XChartImpl.saveAsPNG(XChartImpl.java:114) ~[gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.api.PicFile.save(PicFile.java:207) ~[gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.api.PicFile.createIMG(PicFile.java:176) ~[gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.pages.HtmlPage.addChart(HtmlPage.java:206) [gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.pages.HtmlPage.addChart(HtmlPage.java:213) [gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.EnrichmentReports.createReport(EnrichmentReports.java:813) [gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.EnrichmentReports.createReport(EnrichmentReports.java:633) [gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.EnrichmentReports.createGseaLikeReport(EnrichmentReports.java:350) [gsea-minimal-4.0.3.jar:?]
	at edu.mit.broad.genome.reports.EnrichmentReports.createGseaLikeReport(EnrichmentReports.java:174) [gsea-minimal-4.0.3.jar:?]
	at xtools.gsea.AbstractGsea2Tool.execute_one_with_reporting(AbstractGsea2Tool.java:163) [gsea-minimal-4.0.3.jar:?]
	at xtools.gsea.Gsea.execute(Gsea.java:160) [gsea-minimal-4.0.3.jar:?]
	at xtools.api.AbstractTool.module_main(AbstractTool.java:434) [gsea-minimal-4.0.3.jar:?]
	at org.genepattern.modules.GseaWrapper.main(GseaWrapper.java:273) [gsea-minimal-4.0.3.jar:?]
	at xapps.gsea.CLI.main(CLI.java:29) [gsea-minimal-4.0.3.jar:?]
```

Any fix or suggestions rather than reverting back to GSEA 3.0?

Thanks.

**Comment by davideby** [comment]
Hi @bounlu,

If you've read the thread above you have the basic background: properties of your data set are causing the GSEA computations to result in Infinite (or NaN) values, usually in the NES calculation.  Obviously there's no way to include these on the mountain plots as those are finite in scope.

It's usually a symptom of a data set with a skewed distribution, often where all the values are strictly positive or strictly negative.  If you'd like to discuss *your particular* data set, I recommend posting questions on our [forum](https://groups.google.com/group/gsea-help).  GitHub Issues is meant for bug reports while the forum is better suited for help requests.

We're still discussing the best way to deal with this among the extended team.  Our proposed fix is to de-emphasize and skip reporting on any gene sets with an Infinite or NaN NES value, while also pointing out the likely "skewed distribution" problem.

In the short term, it's actually not so bad that users see error traces like these since it makes it clear that something went wrong with that particular NES calculation and the report for the corresponding gene set shouldn't be used.  We'd like a better solution in the long term but we're still working that out.

Hope that helps.  Contact us on the forum if you have deeper questions about a particular data set (it wouldn't hurt to reference this thread).

- David

**Comment by davideby** [comment]
Fixed in the 4.2.x releases.  We've added warnings, protections, and adjustments (in certain cases) against these where we could.  Any such values included in the datasets will result in warnings about their use.

It is still possible to generate NaN & Infinite values through metric and scoring computations, but these will be detected and warnings issued.  The plotting will no longer fail as we are now explicitly de-emphasizing such values in the plots.
